test/Dialect/Comb/canonicalization.mlir

```diff
@@ -1508,3 +1508,12 @@ hw.module @twoStateICmp(%arg: i4) -> (cond: i1) {
   %0 = comb.icmp bin eq %c-1_i4, %arg : i4
   hw.output %0 : i1
 }
+
+// https://github.com/llvm/circt/issues/5531
+// CHECK-LABEL: @Issue5531
+hw.module @Issue5531(%arg0: i64, %arg1: i64) -> (out: i32) {
+  // CHECK:  %2 = comb.mul %0, %1 {sv.namehint = "hint"} : i32
```

Comment:
One tricky issue here is that strictly speaking `%2` after canonicalization is not equal to `%3`. Actually it makes sense to me to propagate names even when they are not accurate, but we should also consider deriving names from original namehints, such as `hint_0_to_32` for this example.

Reply:
> One tricky issue here is that strictly speaking `%2` after canonicalization is not equal to `%3`. 

When considering the possibilities of overflows, then technically the same could be said about the previous behavior with `%newop = comb.concat %c0_i32, %mul_32_res {sv.namehint = "hint"} : i32, i32` :D
Also, it is a different operation type altogether that gets the name hint (before being optimized away).
But I agree that it is not perfectly accurate.

The reason that I am bringing this up, is that we (@jopperm) use some custom annotations for some `comb` operations in our out-of-tree workflow, and it would be great if they would still exist after `-canonicalize` :)

Reply:
Yeah, preserving optional attributes is hard problem and MLIR doesn't provide a good solution for it (Links for previous discussion on forums: https://discourse.llvm.org/t/canonicalization-passes-dont-keep-attributes/59750/2). Currently we treat sv.namhint specially, and other attributes are not propagated and I think it would be unsafe to propagate optional attributes freely.

Reply:
> Yeah, preserving optional attributes is hard problem and MLIR doesn't provide a good solution for it 

Too bad, anyway, thanks for the link.

> other optional attributes are not propagated at all.

Is this a `comb` specific decision? At least in other dialects, I have seen multiple occasions with `newOp->setDialectAttrs(op->getDialectAttrs());` in their lowering passes.

> I think it would be unsafe to propagate optional attributes freely.

Agreed, this is very much dependent on the intended semantics of the attributes.

Reply:
@uenoku In general, dialect attributes should be propagated best effort, so I think it's appropriate to propagated them through the canonicalizers where it's obvious what the destination/source op is. If that's unsafe in some way, it's the dialect's responsibility to detect that situation, IMO.

Reply:
>  Is this a comb specific decision? At least in other dialects, I have seen multiple occasions with newOp->setDialectAttrs(op->getDialectAttrs()); in their lowering passes.

Yeah, good point. I kind of feel it's generally safe to use `newOp->setDialectAttrs(op->getDialectAttrs())` in lowering passes, since in that case there is some direct mapping from source to target operations (e.g. seq.reg -> sv.reg, hwarith.add -> comb.add). But I feel it's too aggressive to propagate all unknown attributes in canonicalizers since I guess attributes could be moved to random operations.  

> In general, dialect attributes should be propagated best effort, so I think it's appropriate to propagated them through the canonicalizers where it's obvious what the destination/source op is. If that's unsafe in some way, it's the dialect's responsibility to detect that situation, IMO.

I agree that it's ok to propagate attributes when the dest/src are clear.

Reply:
Yeah, all attributes is definitely not safe. That's what dialect attributes are for.

Reply:
> I agree that it's ok to propagate attributes when the dest/src are clear.

So it should be fine here? Both `op` and  `newOp` are of the template type `OpTy`.

Reply:
Yeah, probably. It's still a bit concerning that types are different though. Any thought? @darthscsi